modin/backends/pandas/query_compiler.py

```diff
@@ -1113,39 +1122,54 @@ def map_func(df):
+                ]
+
+            new_index = compute_axis(index) if consider_index else index
+            new_columns = compute_axis(columns) if consider_columns else columns
+            return pandas.MultiIndex.from_product([*new_columns, *new_index])
+
+        if is_all_multi_list and is_correct_len:
             result = result.sort_index()
             index_level_values = [lvl for lvl in obj.index.levels]
-            columns_level_values = [
-                real_cols_bkp.get_level_values(lvl).unique()
-                for lvl in np.arange(real_cols_bkp.nlevels)
-            ]
-            result.index = pandas.MultiIndex.from_product(
-                [*columns_level_values, *index_level_values]
+
+            result.index = compute_index(
+                index_level_values, real_cols_bkp, consider_index=False
             )
             return result
 
         if need_reindex:
-            if isinstance(self.index, pandas.MultiIndex):
-                index_level_values = [
-                    self.index.get_level_values(lvl).unique()
-                    for lvl in np.arange(self.index.nlevels)
-                ]
-                new_index = pandas.MultiIndex.from_product(
-                    [self.columns, *index_level_values]
-                )
+            if is_correct_len:
+                if isinstance(self.index, pandas.MultiIndex):
+                    new_index = compute_index(
+                        self.index, [self.columns], consider_columns=False
+                    )
+                else:
+                    if isinstance(self.columns, pandas.MultiIndex):
+                        new_index = compute_index(
+                            [self.index], self.columns, consider_index=False
+                        )
+                    else:
+                        new_index = compute_index(
+                            [self.index],
+                            [self.columns],
+                            consider_index=False,
+                            consider_columns=False,
+                        )
             else:
                 if isinstance(self.columns, pandas.MultiIndex):
-                    columns_level_values = [
-                        self.columns.get_level_values(lvl).unique()
-                        for lvl in np.arange(self.columns.nlevels)
-                    ]
-                    new_index = pandas.MultiIndex.from_product(
-                        [*columns_level_values, self.index]
-                    )
+                    new_index = compute_index(self.index, self.columns)
                 else:
-                    new_index = pandas.MultiIndex.from_product(
-                        [self.columns, self.index]
+                    new_index = pandas.MultiIndex.from_tuples(
+                        [
+                            tuple([col] + list(tup))
+                            for col in self.columns
+                            for tup in self.index
+                        ]
```

Comment:
I think we can get rid of using slow python `N^2` loop by repeating index and columns and then creating new multi-index from arrays:

Pseudo-code (may be incorrect):
```python
index = self.index.repeat(len(self.columns))
columns = self.columns.repeat(len(self.index))

index_lvls = [x.get_level_values(i) for i in range(index.nlevels)]

new_index = pandas.MultiIndex.from_arrays([columns] + index_lvls)
```

Asymptotic complexity probably still be about `N^2` but in that approach, we will have that loops covered into a libraries with C backend, where evaluating huge loops is faster (i hope so).

Can you please do a time measurement of creating `new_index` with python loops and with using libraries functions?

Reply:
@dchigarev I investigated 3 approaches for computing `new_index`. Using `from_arrays` give ~2x speedup. Table and script for measurements are presented below:

|NROWS x NCOLS|Time approach 1, s|Time approach 2, s|Time approach 3, s|
|-------------------|---------------------|--------------------|----------------------|
|5000 x 2| 0.05 | 0.03 | 0.02|
|5000 x 50| 0.57| 0.33|0.15|
|5000 x 100| 0.85| 0.66|0.27|
|100000 x 100| 13.06| 13.63|6.7|

<details>
<summary>Script</summary>

```python
import modin.pandas as pd
import pandas
import numpy as np
from timeit import default_timer as timer

NCOLS = 100
NROWS = 5000


def generate_columns(ncols):
    return pandas.Index([f"Col{i}" for i in range(ncols)])


def generate_index(nrows):
    l0 = ["a"] * (nrows // 2) + ["b"] * (nrows // 2)
    l1 = np.arange(0, nrows)
    l2 = np.arange(nrows, 2 * nrows)
    l3 = np.arange(2 * nrows, 3 * nrows)
    return pandas.MultiIndex.from_tuples(
        [(i, j, k, l) for i, j, k, l in zip(l0, l1, l2, l3)]
    )


# Calculate new multiindex using from_tuples
def gen_new_multiindex1(index, columns):
    return pandas.MultiIndex.from_tuples(
        [tuple([col] + list(tup)) for col in columns for tup in index]
    )


# Calculate new multiindex using from_tuples using `repeat`
def gen_new_multiindex2(index, columns):
    cols = columns.repeat(len(index))
    idx = list(index) * len(columns)
    return pandas.MultiIndex.from_tuples(
        [tuple([col] + list(tup)) for (col, tup) in zip(cols, idx)]
    )


# Calculate new multiindex using from_arrays
def gen_new_multiindex3(index, columns):
    idx = pandas.MultiIndex.from_tuples(list(index) * len(columns))
    cols = columns.repeat(len(index))
    idx_lvls = [idx.get_level_values(i) for i in range(idx.nlevels)]
    return pandas.MultiIndex.from_arrays([cols] + idx_lvls)


if __name__ == "__main__":
    columns = generate_columns(NCOLS)
    index = generate_index(NROWS)

    t1 = timer()
    res = repr(gen_new_multiindex1(index, columns))
    print("Approach #1:", "{:.2f}".format(timer() - t1), "s")

    t1 = timer()
    res = repr(gen_new_multiindex2(index, columns))
    print("Approach #2:", "{:.2f}".format(timer() - t1), "s")

    t1 = timer()
    res = repr(gen_new_multiindex3(index, columns))
    print("Approach #3:", "{:.2f}".format(timer() - t1), "s")
```
</details>

Approach 3 is used now.

Reply:
how often we will be falling at the if-else branch when we must to compute `new_index` this way? And what is the `pandas.unstack` performance in these cases?

Reply:
This code-block will be used in following conditions:
1. `index` is `MultiIndex`
2. `columns` is `Index`
3. `unstack` was called for all levels (`unstack(level[0,..,nlevels-1])`)
4. `index` if `DataFrame` has non-tree style. I mean next style:
```
           Col0  Col1 
a 0 10 20     0     0
  1 11 21     1     1 
  2 12 22     2     2 
  3 13 23     3     3
b 5 15 25     5     5
  6 16 26     6     6 
  7 17 27     7     7
  8 18 28     8     8 
```

instead of tree-style:
```
           Col0  Col1 
a  X   10     0     0
       20     1     1 
   Y   10     2     2 
       20     3     3
b  X   10     5     5
       20     6     6 
   Y   10     7     7
       20     8     8 
```

In this case `pandas.DataFrame.unstack` is faster to ~100 times